daemon/socket.c

```diff
@@ -320,7 +325,28 @@ static ssize_t __ip_recvfrom_ts(socket_t *s, void *buf, size_t len, endpoint_t *
 			if (cm->cmsg_level == SOL_SOCKET && cm->cmsg_type == SO_TIMESTAMP) {
 				*tv = *((struct timeval *) CMSG_DATA(cm));
 				tv = NULL;
-				break;
+			} else if (cm->cmsg_level == IPPROTO_IP && cm->cmsg_type == IP_PKTINFO) {
+				struct in_pktinfo *pktinfo = (struct in_pktinfo *)CMSG_DATA(cm);
+
+				if (is_addr_unspecified(&s->local.address))
+					ilog(LOG_WARNING, "Socket local address is already specified");
+
+				s->local.address.u.ipv4 = pktinfo->ipi_spec_dst;
```

Comment:
There's a potential race condition here as multiple threads might be accessing the s->local struct simultaneously.  Originally this was supposed to be filled during socket creating and then be read-only afterwards, making thread synchronization unnecessary.